tests/testthat/test-qenv-within.R

```diff
@@ -0,0 +1,117 @@
+# styler: off
+# nolint start
+
+# code acceptance ----
+testthat::test_that("simple and compound expressions are evaluated", {
+  q <- new_qenv()
+  testthat::expect_no_error(
+    within(q, 1 + 1)
+  )
+  testthat::expect_no_error(
+    within(q, {
+      1 + 1
+    })
+  )
+})
+
+# code identity ----
+testthat::test_that("styling of input code does not impact evaluation results", {
+  q <- new_qenv()
+  q <- within(q, 1 + 1)
+  q <- within(q, {1 + 1})
+  q <- within(q, {
+    1 + 1
+  })
+  q <- within(q, {
+    1 +
+      1
+  })
+  all_code <- get_code(q)
+  testthat::expect_identical(
+    all_code,
+    rep("1 + 1", 4L)
+  )
+
+  q <- new_qenv()
+  q <- within(q, {1 + 1; 2 + 2})
+  q <- within(q, {
+    1 + 1; 2 + 2
+  })
+  q <- within(q, {
+    1 + 1
+    2 + 2
+  })
+  q <- within(q, {
+    1 + 1;
+    2 + 2
+  })
+  all_code <- get_code(q)
+  testthat::expect_identical(
+    all_code,
+    rep(c("1 + 1", "2 + 2"), 4L)
+  )
+})
+
+
+# return value ----
+testthat::test_that("within.qenv renturns a deep copy of `data`", {
+  q <- new_qenv()
+  q <- within(new_qenv(), i <- iris)
+  qq <- within(q, {})
+  testthat::expect_equal(q, qq)
+  testthat::expect_false(identical(q, qq))
+})
+
+testthat::test_that("within.qenv renturns qenv.error even if evaluation raises error", {
+  q <- new_qenv()
+  q <- within(q, i <- iris)
+  qq <- within(q, stop("right there"))
+  testthat::expect_true(
+    exists("qq", inherits = FALSE)
+  )
+  testthat::expect_s3_class(qq, "qenv.error")
+})
+
+
+# injecting values ----
+testthat::test_that("external values can be injected into expressions through `...`", {
+  q <- new_qenv()
+
+  q1 <- within(q, {
+    i <- subset(iris, Species == "virginica")
+  })
+
+  q2 <- within(q, {
+    i <- subset(iris, Species == species)
+  },
+  species = "virginica")
+
+  external_value <- "virginica"
+  q3 <- within(q, {
+    i <- subset(iris, Species == species)
+  },
+  species = external_value)
+
+  testthat::expect_identical(get_code(q1), get_code(q2))
+  testthat::expect_identical(get_code(q1), get_code(q3))
```

Comment:
If the test reflected its description, then it would look as suggested. It would be shorter and more precise


```suggestion
  external_value <- "virginica"
  q <- within(
    new_qenv(), 
    i <- subset(iris, Species == species),
    species = external_value
   )

  testthat::expect_identical(get_code(q), 'i <- subset(iris, Species == "virginica")')
```

Reply:
Isn't the `q2` case worth testing as well?

Reply:
I don't think so. If we consider `q2` important, then we can also find another case. For example:

```r
foo <- function() "virginica"
within(iris$Species == species, species = foo())
```